src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs

```diff
@@ -0,0 +1,86 @@
+                    .Concat(base.GetCompletions(context, environmentSettings));
+            }
+            else
+            {
+                //no exact match on short name
+                if (!templateGroups.Any(template => template.ShortNames.Contains(args.ShortName)))
+                {
+                    return templateGroups
+                        .SelectMany(g => g.ShortNames, (g, shortName) => new CompletionItem(shortName, documentation: g.Description))
+                        .Where(c => c.Label.StartsWith(args.ShortName))
+                        .Distinct()
+                        .OrderBy(c => c.Label, StringComparer.OrdinalIgnoreCase)
+                        .Concat(base.GetCompletions(context, environmentSettings));
+                }
+
+                //if there is exact match do further reparsing
+                HashSet<CompletionItem> distinctCompletions = new HashSet<CompletionItem>();
+                foreach (TemplateGroup templateGroup in templateGroups.Where(template => template.ShortNames.Contains(args.ShortName)))
+                {
+                    foreach (IGrouping<int, CliTemplateInfo> templateGrouping in templateGroup.Templates.GroupBy(g => g.Precedence).OrderByDescending(g => g.Key))
+                    {
+                        foreach (CliTemplateInfo template in templateGrouping)
+                        {
+                            TemplateCommand command = new TemplateCommand(this, environmentSettings, templatePackageManager, templateGroup, template);
+                            Parser parser = ParserFactory.CreateParser(command);
+                            ParseResult templateParseResult = parser.Parse(context.CommandLineText);
+                            foreach (CompletionItem completion in templateParseResult.GetCompletions(context.CursorPosition))
+                            {
+                                distinctCompletions.Add(completion);
+                            }
+                        }
+                    }
+                }
+                return distinctCompletions.OrderBy(c => c.Label, StringComparer.OrdinalIgnoreCase);
+            }
+        }
+
+        protected internal override IEnumerable<CompletionItem> GetCompletions(CompletionContext context, IEngineEnvironmentSettings environmentSettings)
+        {
+            if (context is not TextCompletionContext textCompletionContext)
+            {
+                return base.GetCompletions(context, environmentSettings);
+            }
+
+            InstantiateCommandArgs args = ParseContext(context.ParseResult);
+
+            using TemplatePackageManager templatePackageManager = new TemplatePackageManager(environmentSettings);
+            HostSpecificDataLoader? hostSpecificDataLoader = new HostSpecificDataLoader(environmentSettings);
+
+            //TODO: consider new API to get templates only from cache (non async)
+            IReadOnlyList<ITemplateInfo> templates =
```

Comment:
https://github.com/dotnet/templating/issues/2191:  consider creating TemplatePackageManager.GetTemplates version which uses cache only for tab completion and help